src/group_theory/nilpotent.lean

```diff
@@ -0,0 +1,285 @@
+the associated quotient of `G`. This means that if `G` is not nilpotent, then
+none of what we have called `upper_central_series G`, `lower_central_series G` or
+the sequences satisfying `is_ascending_central_series` or `is_descending_central_series`
+are actually central series. Note that the fact that the upper and lower central series
+are not central series if `G` is not nilpotent is a standard abuse of notation.
+
+-/
+
+open subgroup
+
+variables {G : Type*} [group G] (H : subgroup G) [normal H]
+
+/-- If `H` is a normal subgroup of `G`, then the set `{x : G | ∀ y : G, x*y*x⁻¹*y⁻¹ ∈ H}`
+is a subgroup of `G` (because it is the preimage in `G` of the centre of the
+quotient group `G/H`.)
+-/
+def upper_central_series_step : subgroup G :=
+{ carrier := {x : G | ∀ y : G, x * y * x⁻¹ * y⁻¹ ∈ H},
+  one_mem' := λ y, by simp [subgroup.one_mem],
+  mul_mem' := λ a b ha hb y, begin
+    convert subgroup.mul_mem _ (ha (b * y * b⁻¹)) (hb y) using 1,
+    group,
+  end,
+  inv_mem' := λ x hx y, begin
+    specialize hx y⁻¹,
+    rw [mul_assoc, inv_inv] at ⊢ hx,
+    exact subgroup.normal.mem_comm infer_instance hx,
+  end }
+
+lemma mem_upper_central_series_step (x : G) :
+  x ∈ upper_central_series_step H ↔ ∀ y, x * y * x⁻¹ * y⁻¹ ∈ H := iff.rfl
+
+open quotient_group
+
+/-- The proof that `upper_central_series_step H` is the preimage of the centre of `G/H` under
+the canonical surjection. -/
+lemma upper_central_series_step_eq_comap_center :
+  upper_central_series_step H = subgroup.comap (mk' H) (center (quotient H)) :=
+begin
+  ext,
+  rw [mem_comap, mem_center_iff, forall_coe],
+  apply forall_congr,
+  intro y,
+  change x * y * x⁻¹ * y⁻¹ ∈ H ↔ ((y * x : G) : quotient H) = (x * y : G),
+  rw [eq_comm, eq_iff_div_mem, div_eq_mul_inv],
+  congr' 2,
+  group,
+end
+
+instance : normal (upper_central_series_step H) :=
+begin
+  rw upper_central_series_step_eq_comap_center,
+  apply_instance,
+end
+
+variable (G)
+
+/-- An auxiliary type-theoretic definition defining both the upper central series of
+a group, and a proof that it is normal, all in one go. -/
+def upper_central_series_aux : ℕ → Σ' (H : subgroup G), normal H
+| 0 := ⟨⊥, infer_instance⟩
+| (n + 1) := let un := upper_central_series_aux n, un_normal := un.2 in
+   by exactI ⟨upper_central_series_step un.1, infer_instance⟩
+
+/-- `upper_central_series G n` is the `n`th term in the upper central series of `G`. -/
+def upper_central_series (n : ℕ) : subgroup G := (upper_central_series_aux G n).1
+
+instance (n : ℕ) : normal (upper_central_series G n) := (upper_central_series_aux G n).2
+
+lemma upper_central_series_zero_def : upper_central_series G 0 = ⊥ := rfl
+
+/-- The `n+1`st term of the upper central series `H i` has underlying set equal to the `x` such
+that `⁅x,G⁆ ⊆ H n`-/
+lemma mem_upper_central_series_succ_iff {G : Type*} [group G] (n : ℕ) (x : G) :
+  x ∈ upper_central_series G (n + 1) ↔
+  ∀ y : G, x * y * x⁻¹ * y⁻¹ ∈ upper_central_series G n := iff.rfl
+
```

Comment:
Would it make sense to make `upper_central_series` `irreducible` at this point, so `mem_upper_central_series_succ_iff` is the only thing you're allowed to know?

Reply:
If this happens then some later proofs break, e.g. 

```
lemma upper_central_series_is_ascending_central_series :
  is_ascending_central_series (upper_central_series G) :=
⟨rfl, λ x n h, h⟩
```
 
neither proof works any more because both proofs abuse definitional equality. Of course I can fix these but right now this short proof is going to become some rewrites if I do it. I could just make it irreducible at the end, but that's like saying "I abused defeq but I don't want you to". So I will not make this change but of course I won't complain if someone has a strong opinion that this is what we should be doing.